daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/ConcurrentCompiledPackages.scala

```diff
@@ -83,27 +84,28 @@ final class ConcurrentCompiledPackages extends MutableCompiledPackages {
           // map using 'computeIfAbsent' which will ensure we only compile the
           // package once. Other concurrent calls to add this package will block
           // waiting for the first one to finish.
-          _packages.computeIfAbsent(
-            pkgId, { _ =>
-              // Compile the speedy definitions for this package.
-              val defns =
-                speedy
-                  .Compiler(packages orElse state.packages, stackTraceMode, profilingMode)
-                  .unsafeCompilePackage(pkgId)
-              defns.foreach {
-                case (defnId, defn) => _defns.put(defnId, defn)
-              }
-              // Compute the transitive dependencies of the new package. Since we are adding
-              // packages in dependency order we can just union the dependencies of the
-              // direct dependencies to get the complete transitive dependencies.
-              val deps = pkg.directDeps.foldLeft(pkg.directDeps) {
-                case (deps, dependency) =>
-                  deps union _packageDeps.get(dependency)
-              }
-              _packageDeps.put(pkgId, deps)
-              pkg
+          if (!_packages.contains(pkgId)) {
```

Comment:
Cannot you use `putIfAbsent`

Reply:
Isn’t `putIfAbsent` strict in the value and will therefore compute all the other stuff even if don’t end up putting?

Reply:
Yeah, unfortunately. Java's `Map` has `computeIfAbsent`, which is lazy, but there doesn't seem to be an equivalent in Scala.

Reply:
Oh, I see the original code used just that.

Reply:
Could we  use `getOrElseUpdate` in this case ?